Please support ''file' URL scheme

To get around issues with proxies, credentials etc,
I download some p2 repos and install locally.

old p2AsMaven supported file urls for repos.

this new p2deps does not :-(

```
Caused by: java.lang.IllegalArgumentException: Expected URL scheme 'http' or 'https' but was 'file'
	at okhttp3.HttpUrl$Builder.parse$okhttp(HttpUrl.kt:1254)
	at okhttp3.HttpUrl$Companion.get(HttpUrl.kt:1634)
	at okhttp3.Request$Builder.url(Request.kt:184)
	at dev.equo.solstice.p2.P2Client.buildRequest(P2Client.java:146)
```

--- a/solstice/src/main/java/dev/equo/solstice/p2/JarCache.java
+++ b/solstice/src/main/java/dev/equo/solstice/p2/JarCache.java
@@ -15,7 +15,11 @@ package dev.equo.solstice.p2;
 
 import java.io.File;
 import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.nio.file.Files;
+import java.nio.file.Path;
+
 import okhttp3.OkHttpClient;
 import okio.Okio;
 
@@ -35,7 +39,14 @@ class JarCache {
 		if (jar.isFile()) {
 			return jar;
 		}
-		if (cachingPolicy.networkAllowed()) {
+		if (P2Client.isFileUrl(unit.getJarUrl())) {
+			try {
+				Files.copy(Path.of(new URI(unit.getJarUrl())), jar.toPath());
+			} catch (URISyntaxException e) {
+				throw new IllegalArgumentException("Cannot parse url", e);
+			}
+			return jar;
+		} else if (cachingPolicy.networkAllowed()) {
 			if (!repoDir.isDirectory()) {
 				FileMisc.mkdirs(repoDir);
 				FileMisc.writeToken(repoDir, ".url", unit.getRepoUrl());

--- a/solstice/src/main/java/dev/equo/solstice/p2/P2Client.java
+++ b/solstice/src/main/java/dev/equo/solstice/p2/P2Client.java
@@ -16,9 +16,14 @@ package dev.equo.solstice.p2;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.UnknownHostException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -151,6 +156,10 @@ public class P2Client implements AutoCloseable {
 		return builder.build();
 	}
 
+	static boolean isFileUrl(String url) {
+		return url.startsWith("file://");
+	}
+
 	private byte[] getBytes(String url) throws IOException, NotFoundException {
 		if (cachingPolicy.tryOfflineFirst()) {
 			var cached = offlineMetadataCache.get(url);
@@ -161,7 +170,18 @@ public class P2Client implements AutoCloseable {
 				return cached;
 			}
 		}
-		if (cachingPolicy.networkAllowed()) {
+		if (isFileUrl(url)) {
+			try {
+				return Files.readAllBytes(Path.of(new URI(url)));
+			} catch (URISyntaxException e) {
+				throw new IOException("Cannot parse file url", e);
+			} catch (NoSuchFileException e) {
+				if (cachingPolicy.cacheAllowed()) {
+					offlineMetadataCache.put404(url);
+				}
+				throw new NotFoundException(url);
+			}
+		} else if (cachingPolicy.networkAllowed()) {
 			var request = buildRequest(url);
 			try (var response = metadataClient.newCall(request).execute()) {
 				if (response.code() == 404) {

